src/main/java/net/eldeen/dropwizard/CfSignalResourceBundle.java

```diff
@@ -7,75 +17,154 @@
+
+      if (amazonCloudFormation != null) {
+        return amazonCloudFormation;
+      }
+
+      return cloudFormationAtomicRef.updateAndGet((unused) -> {
+
+        AmazonCloudFormationClient amazonCloudFormationClient = new AmazonCloudFormationClient();
+
+        amazonCloudFormationClient.setRegion(Region.getRegion(Regions.valueOf(cfSignalResourceConfig.getAwsRegion()
+                                                                                                    .toUpperCase()
+                                                                                                    .replace('-', '_'))));
+        return amazonCloudFormationClient;
+      });
+    };
+  }
+
+  @Inject
+  public CfSignalResourceBundle(@CfSignalResourceInstanceId Optional<String> instanceId, AmazonCloudFormation amazonCloudFormation) {
+    instanceIdProvider = () -> checkNotNull(instanceId);
+    cloudFormationSupplier = (config) -> amazonCloudFormation;
   }
 
   @Override
   public void initialize(Bootstrap<?> bootstrap) {
+  }
 
+  public CfSignalResourceConfig getConfiguration() {
+    return new CfSignalResourceConfig();
   }
 
   @Override
-  public void run(Environment environment) {
+  public void run(T config, Environment environment) {
 
-    final String instanceId = EC2MetadataUtils.getInstanceId();
-    if (instanceId == null || instanceId.trim() == "") {
+    final Optional<String> instanceId = instanceIdProvider.get();
+    if (!instanceId.isPresent()) {
       LOGGER.debug("Unable to fetch EC2 Instance ID, assuming not running on AWS and thus not signalling");
       return;
     }
 
-    environment.lifecycle().addServerLifecycleListener(new ServerLifecycleListener() {
-      @Override
-      public void serverStarted(Server server) {
-        AmazonCloudFormation client = null;
+    environment.lifecycle()
+               .addLifeCycleListener(
+                 new CfSignalResourceLifcycleListener(getCfResourceBundleConfig(config).orElseGet(this::getConfiguration),
+                                                      instanceId.get()));
+  }
+
+  private void sendSignal(CfSignalResourceConfig config, final String instanceId, boolean success) {
+    try {
+      AmazonCloudFormation client = cloudFormationSupplier.apply(config);
+      SignalResourceRequest request = new SignalResourceRequest();
+      request.setUniqueId(instanceId);
+      request.setLogicalResourceId(config.getAsgResourceName());
+      request.setStackName(config.getStackName());
+      request.setStatus(success? ResourceSignalStatus.SUCCESS : ResourceSignalStatus.FAILURE);
+      client.signalResource(request);
+    }
+    catch (Exception e) {
+      LOGGER.error("There was a problem signaling " + config.getAsgResourceName()
+                   + " in stack " + config.getStackName(), e);
+    }
+    finally {
+      AmazonCloudFormation internalClient = cloudFormationAtomicRef.get();
+      if (internalClient != null) {
+        internalClient.shutdown();
+      }
```

Comment:
One more thing: with the function design, it's not possible to test client shutdown.